### ISSUE TITLE
k-fold cross validation for training models with CLI

### ISSUE THREAD

**Issue by danielduckworth** [uber/ludwig#462]
**Describe the use case**
Training multiple models with shuffled training data (k-folds) can reveals information about our data sets. For example, one fold may be much less accurate than other folds telling us that we might need to increase the total number of samples.

**Describe the solution you'd like**
For the CLI, it would be useful if we could have a parameter to train with k-fold cross validation.

**Describe alternatives you've considered**
k-fold can be achieved with the python API but not as convenient as the CLI.

**Additional context**
I can provide the sample code I've been using for the API.

**Comment by msaisumanth** [comment]
@danielduckworth Thanks for the suggestion. This is a good idea. 

Feel free to create a pr if you have time to work on it :)

**Comment by danielduckworth** [comment]
@msaisumanth I'll work on it with my team and submit something in the next few weeks. We have already benefited greatly from Ludwig so happy to contribute.

**Comment by w4nderlust** [comment]
Thank you so much @danielduckworth . I would suggest you do do a draft minimal not working pr so that we can give you guidance on the design first, so that when you end up implementing it it's more likely there won't be any misalignment.

**Comment by the-learner** [comment]
Hello @w4nderlust , I am working with @danielduckworth on this k-fold contrib. I want to know what do you expect as the structure of the minimal draft over here. Should I right it as a standalone function that will be imported or should it be added as an extra parameter in the full_train() function and add the parameter in cli() function. Sorry if I am asking something obvious but this is my first time contributing :)

**Comment by msaisumanth** [comment]
@the-learner Thanks a lot for working on this! I think what @w4nderlust was suggesting was to outline the high level flow so that we're all on the same page as to how to implement it. 

You can write the logic in full_train first and then adding the cli stuff should be fairly easy. So please tell us how you would implement this in full_train at a high level and then you can start coding. If you think it's straightforward, go right ahead and we can review the code.

**Comment by the-learner** [comment]
@msaisumanth The change seems very straight forward but I will tell high level approach anyway.

K-fold k value will be taken input into full_train. There will be different combinations of data (k precisely), based on those data combinations, the training will be done. I might not have to change much of a functionality in the main  training function as well.

**Comment by w4nderlust** [comment]
@the-learner there are a bunch of additional considerations: if you have more than one split, the HDF5 and JSON files will be saved for all the splits? Will they be skipped? What about the results? Will you have k different results directories or just one containing all 3 of them? And what about aggregate results (with mean and standard deviation)?

**Comment by the-learner** [comment]
@w4nderlust That was indeed a question on my mind for some time. I think the function of k-fold splits should be to let the user know that the data size needs to be increased if the accuracy values difference will be considerably large. Because saving aggregate results, or most accurate results will just be an accuracy hack. I think it will just store models and final results for one of the splits, but functionality of k-fold can be to inform user about the nature of data. So the results on the test data will be displayed for the k-splits but final model should just be one as it is now.

**Comment by danielduckworth** [comment]
I've discussed this with @the-learner and I think what makes the most sense is to stay in keeping with the way ludwig already works. So that means storing separate HDF5 and JSON files for each split.

The results of each k-fold would also be saved separately so they can be reported for each fold and the mean calculated and reported.

I think the final step would then be to train using all the data (no split) as the final model. The user can then decide on the basis of the k-fold results whether they want to use the final model.

I'm not sure about the convention for the directories. Do you have any preference @w4nderlust?

**Comment by w4nderlust** [comment]
This makes sense. A few more considerations:
- in case the k fold is specified I would train just k models and not a final model on all the splits, as that can always be done anyway by calling ludwig again without the split parameter
- when ludwig creates the HDF5 and JSON files, they are placed in the same directory of the original data with the same name, so in order to make the k fold work, you will probably have to either name the files created for each for differently (there's no way to doing it at the moment) or not saving HDF5 and JSON at all
- code-wise, the way I'm thinking about this is a `kfold_cross_validate()` function in `train..py` that calls `full_train()` `k` times. In the main `cli()` function there would be an if that checks if a `kfold` command line parameter is psecified, if not, `full_train` will be called, otherwise `kfold_cross_validate` will be called. But I'm also open to other designs if you think there's a better way.
- after all the k fold models have been trained and the results have been obtained, I would create either an additional JSON file containing the k fold summarized statistics (something like `kfold_training_statistics.json`) so that we could later also build visualizations, like violin plots of the like
- as the different k fold training processes are entirely independent of each other, it would be interesting to find a way to run them in parallel on separate processes. The problem I see here is the fact that, in particular when using GPUs, TensorFlow tries to grab all the resources possible, so running multiple things on the same GPU is kinda tricky, you would have to reduce the percentage of ram used for each process for instance, but if you have as many GPUs as folds, that would be pretty easy to do. Anyway, this could be done also later on, but I wanted to put it here as I believe it would be pretty useful.

**Comment by jimthompson5802** [comment]
@danielduckworth @the-learner @w4nderlust @msaisumanth 

From what I can tell there has been no activity on this issue since August.  If there is still an interest in providing this capability, I have a working prototype based on guidance given in the last response.  If there is an interest, I'll submit a PR for this prototype to continue the development effort.

I followed these points:
```
code-wise, the way I'm thinking about this is a kfold_cross_validate() function in 
train..py that calls full_train() k times. In the main cli() function there would be an if that 
checks if a kfold command line parameter is psecified, if not, full_train will be called, 
otherwise kfold_cross_validate will be called. But I'm also open to other designs if you 
think there's a better way.

after all the k fold models have been trained and the results have been obtained, I would 
create either an additional JSON file containing the k fold summarized statistics 
(something like kfold_training_statistics.json) so that we could later also build visualizations, 
like violin plots of the like
```
A new parameter `--k_fold` is now available for the `ludwig train` command.  This is the help output for the new parameter:
```
 -kf K_FOLD, --k_fold K_FOLD
                        number of folds for a k-fold cross validation run
```

Here is a sample execution with the enhancement:
```
ludwig train \
  --model_definition_file=model1_definition.yaml \
  --data_csv=data/train.csv \
  --output_directory=results \
  --k_fold=3
```
Model definition and data used for the above is based on the Titanic survivor example.
This is the output from the above example:
[k-fold_test_output.txt](https://github.com/uber/ludwig/files/4006359/k-fold_test_output.txt)

Training statistics for each fold is consolidated into a single json file that is saved in the location pointed to by the `output_directory` parameter.  The individual training statistics are attached to a high-level key in the consolidated file, e.g.,
```
{
  "fold_1': {training statistics from fold 1},
  "fold_2": {training statistics from fold 2},
  "fold_3": {training statistics from fold 3}
}
```
This is the consolidated statistics from the sample run
[kfold_training_statistics_json.txt](https://github.com/uber/ludwig/files/4006376/kfold_training_statistics_json.txt)

To handle the naming issue, this prototype creates temporary directory to hold training and validation data sets created by performing the k-fold splitting of the original training data.  This is not the most efficient implementation since preprocessing on the raw training data is repeated `k` times.  I took this approach because it seemed the fastest way to get a working prototype.

**Comment by w4nderlust** [comment]
@jimthompson5802  thank you! This actually sounds really promising! Please go ahead with the PR so that we cna look at it. I understand the problem with preprocessing data 3 times, but I would rather have the kfold functionality with this limitation (that can be fixed later) rather than not having it.
One final thing: in the kfold statistics file, maybe there could be an additional `'aggregate': {...}` field that aggregates the overall stats, for instance it it's a classification task it may contain the mean accuracy of all the folds and the standard deviation of that accuracy, and the same for all other metrics (not the per-class ones).

**Comment by w4nderlust** [comment]
The great  PR by @jimthompson5802 #608 is now merged, we can close this.

**Comment by w4nderlust** [comment]
If anyone wants to help on surfacing this feature also in the API feel free to open a new issue.

**Comment by muhammad-yasir** [comment]
Hi there, 
I am using ludwig 0.3.3. I am trying to train the model using --k_fold option. I have two issues:

1) When I run the following command, it gives me  this error (ludwig train: error: unrecognized arguments: --k_fold=10):
ludwig train --dataset train.csv --config_file model_definition.yml --output_directory results --k_fold=10

2) Secondly, Please clear my confusion.  If I am not mistaken, in traditional learning, when we use k-fold validation, we first instantiate the object of the learning model and give the k_fold value as an argument and then we split the training and test data. In Ludwig, do we first split the training and test data into a separate file and then during training we mentioned the k-fold value and train the model and later predict the test data using the learned model. Am I right? or am I missing something?

**Comment by jimthompson5802** [comment]
@muhammad-yasir Thank you for submitting your questions.

**re: #1**.  
With v0.3 some breaking changes were introduced that affected cli parameters.  The k-fold functionality is now part of the ludwig `experiment` function.  Please refer to the [updated user guide for details on how to invoke k-fold cv](https://ludwig-ai.github.io/ludwig-docs/user_guide/#experiment).

If the above advice does not solve the problem, please open a new issue that we can use to track the problem.  When opening a new issue, please provide the full command used and the generated log output.

**re: #2**. 
 For a given model configuration specification, the scope of Ludwig's k-fold function is to provide the user with more robust understanding the model performance as reported in the k-fold cv statistics data captured in the json output.  To do this Ludwig internally partitions the data specified by the `dataset` parameter into the specified folds.

As a good ML practice the user should have set aside a hold out `test` data set separately, irrespective of the use k-fold function, that can be used for the final assessment of model performance.

**Comment by muhammad-yasir** [comment]
Thanks Jim. It worked. Thanks for your positive response.
